storage/storageCacherAdapter/storageCacherAdapter.go

```diff
@@ -0,0 +1,243 @@
+			continue
+		}
+
+		evictedValBytes, err := c.marshalizer.Marshal(evictedVal)
+		if err != nil {
+			log.Error("could not marshall value", "error", err)
+			continue
+		}
+
+		err = c.db.Put([]byte(evictedKeyStr), evictedValBytes)
+		if err != nil {
+			log.Error("could not save to db", "error", err)
+			continue
+		}
+	}
+
+	return len(evictedValues) != 0
+}
+
+// Get returns the value at the given key
+func (c *storageCacherAdapter) Get(key []byte) (value interface{}, ok bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	val, ok := c.cacher.Get(string(key))
+	if ok {
+		return val, true
+	}
+
+	valBytes, err := c.db.Get(key)
+	if err != nil {
+		return nil, false
+	}
+
+	storedData := c.storedDataFactory.CreateEmpty()
+	err = c.marshalizer.Unmarshal(storedData, valBytes)
+	if err != nil {
+		log.Error("could not unmarshall", "error", err)
+		return nil, false
+	}
+
+	return storedData, true
+}
+
+// Has checks if the given key is present in the storageUnit
+func (c *storageCacherAdapter) Has(key []byte) bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	isPresent := c.cacher.Contains(string(key))
+	if isPresent {
+		return true
+	}
+
+	err := c.db.Has(key)
+	if err == nil {
+		return true
+	}
+
+	return false
+}
+
+// Peek returns the value at the given key
+func (c *storageCacherAdapter) Peek(key []byte) (value interface{}, ok bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	return c.cacher.Peek(string(key))
+}
+
+// HasOrAdd adds the given value in the storageUnit
+func (c *storageCacherAdapter) HasOrAdd(key []byte, value interface{}, sizeInBytes int) (has, added bool) {
+	_ = c.Put(key, value, sizeInBytes)
+
+	return false, true
+}
+
+// Remove deletes the given key from the storageUnit
+func (c *storageCacherAdapter) Remove(key []byte) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	removed := c.cacher.Remove(string(key))
+	if removed {
+		return
+	}
+
+	_ = c.db.Remove(key)
+}
+
+// Keys returns all the keys present in the storageUnit
+func (c *storageCacherAdapter) Keys() [][]byte {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	cacherKeys := c.cacher.Keys()
+	storedKeys := make([][]byte, 0)
+	for i := range cacherKeys {
+		key, ok := cacherKeys[i].(string)
+		if !ok {
+			continue
+		}
+
+		storedKeys = append(storedKeys, []byte(key))
+	}
+
+	getKeys := func(key []byte, _ []byte) bool {
+		storedKeys = append(storedKeys, key)
+		return true
+	}
+
+	c.db.RangeKeys(getKeys)
+	return storedKeys
+}
+
+// Len returns the number of elements from the storageUnit
+func (c *storageCacherAdapter) Len() int {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	cacheLen := c.cacher.Len()
+	countValues := func(_ []byte, _ []byte) bool {
+		cacheLen++
+		return true
+	}
+
+	c.db.RangeKeys(countValues)
+	return cacheLen
+}
+
+// SizeInBytesContained returns the number of bytes stored in the storageUnit
+func (c *storageCacherAdapter) SizeInBytesContained() uint64 {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	size := c.cacher.SizeInBytesContained()
```

Comment:
Done. Refactored so that Len() will return the number of values from storage+cacher, but SizeInBytesContained() will return only the num of bytes contained in the cacher. To also compute the number of bytes from storage, for every Remove() from DB, a get() will be needed to see the size of the object contained in the db, and I do not think this is needed.